pkg/logs/input/docker/convertor.go

```diff
@@ -0,0 +1,160 @@
+)
+
+// Convertor specific for docker logs read from socket.
+type Convertor struct {
+	iParser.Convertor
+}
+
+// Convert converts the input message to Line, if it's partial msg, defaultPrefix will be
+// used as prefix.
+func (c *Convertor) Convert(msg []byte, defaultPrefix iParser.Prefix) *iParser.Line {
+	components := bytes.SplitN(msg, delimiter, numOfComponents)
+	if !c.validateHeader(components[indexOfHeader]) {
+		// take this msg as partial log splitted by upstream.
+		return &iParser.Line{
+			Prefix:  defaultPrefix,
+			Content: msg,
+			Size:    len(msg),
+		}
+	}
+
+	if c.emptyContent(components) {
+		return nil
+	}
+
+	status, timestamp, contentLen := c.parseHeader(components[indexOfHeader])
+	if contentLen <= 0 && status == "" { // take it as tty, which means msg does not contain 8 bytes docker header.
+		return &iParser.Line{
+			Prefix: iParser.Prefix{
+				Status:    message.StatusInfo,
+				Timestamp: timestamp,
+			},
+			Content: components[indexOfContent],
+			Size:    len(components[indexOfContent]),
+		}
+	}
+	content := c.parseContent(msg)
+	return &iParser.Line{
+		Prefix: iParser.Prefix{
+			Status:    status,
+			Timestamp: timestamp,
+		},
+		Content: content,
+		Size:    len(content),
+	}
+}
+
+func (c *Convertor) emptyContent(components [][]byte) bool {
+	// consider as empty content when:
+	// there is no content byte array at all, or
+	// content is considered to be empty.
+	return len(components) < numOfComponents || c.isEmptyMsg(components[indexOfContent])
+}
+
+// parseContent removes the 8 byte header, timestamp, and space that occurs between 16KB section of a log.
+// If a docker log is greater than 16KB, each 16KB partial section will have a header, timestamp, and space in front of it.
+// For example, a message that is 35KB will be of the form: `H M1H M2H M3` where "H" is what pre-pends each 16 KB section.
+// This function removes the "H " between two partial messages sections while leaving the very first "H ".
+// Input:
+//   H M1H M2H M3
+// Output:
+//   H M1M2M3
+func (c *Convertor) parseContent(msg []byte) []byte {
+	var result bytes.Buffer
+	for len(msg) > 0 {
+		components := bytes.SplitN(msg, delimiter, numOfComponents)
+		_, _, length := c.parseHeader(components[indexOfHeader])
+		result.Write(components[indexOfContent][:length])
+		msg = components[indexOfContent][length:]
+	}
+	return result.Bytes()
+}
+
+func bytesToUint32(bytes []byte) int {
+	return int(binary.BigEndian.Uint32(bytes))
+}
+
+func (c *Convertor) parseHeader(header []byte) (string, string, int) {
+	var status string
+	var timestamp string
+	var contentLen int
+	// rely on the first byte of header to know whether this header contains leading
+	// 8 Bytes docker header or not.
+	// If header contains leading 8 bytes metadata, the first byte should be the status
+	// with value 1(stdout) or 2(stderr)
+	// otherwise, treat this header as tty log, in which case, the header is just
+	// a timestamp.
+	switch header[indexOfStatus] {
```

Comment:
yes, that will be the case, since we already validate the header+timestamp in the very beginning, I think this assumption here should be fine. What do you think?